rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,393 @@
+core.status.set(
+  core.status.core$.pipe(core => core.elasticsearch);
+)
+```
+
+# Motivation
+
+Kibana should do as much possible to help users keep their installation in a working state. This includes providing as much detail about components that are not working as well as ensuring that failures in one part of the application do not block using other portions of the application.
+
+In order to provide the user with as much detail as possible about any systems that are not working correctly, the status mechanism should provide excellent defaults in terms of expressing relationships between services and presenting detailed information to the user.
+
+# Detailed design
+
+## Failure Guidelines
+
+While this RFC primarily describes how status information is signaled from individual services and plugins to Core, it's first important to define how Core expects these services and plugins to behave in the face of failure more broadly.
+
+Core is designed to be resilient and adaptive to change. When at all possible, Kibana should automatically recover from failure, rather than requiring any kind of intervention by the user or administrator.
+
+Given this goal, Core expects the following from plugins:
+- During initialization, `setup`, and `start` plugins should only throw an exception if a truly unrecoverable issue is encountered. Examples: HTTP port is unavailable, server does not have the appropriate file permissions.
+- Temporary error conditions should always be retried automatically. A user should not have to restart Kibana in order to resolve a problem when avoidable. This means all initialization code should include error handling and automated retries. Examples: creating an Elasticsearch index, connecting to an external service.
+  - It's important to note that some issues do require manual intervention in _other services_ (eg. Elasticsearch). Kibana should still recover without restarting once that external issue is resolved.
+- Unhandled promise rejections are not permitted. In the future, Node.js will crash on unhandled promise rejections. It is impossible for Core to be able to properly handle and retry these situations, so all services and plugins should handle all rejected promises and retry when necessary.
+- Plugins should only crash the Kibana server when absolutely necessary. Some features are considered "mission-critical" to customers and may need to halt Kibana if they are not functioning correctly. Example: audit logging.
+
+## API Design
+
+### Types
+
+```ts
+/**
+ * The current status of a service at a point in time.
+ * 
+ * @typeParam Meta - JSON-serializable object. Plugins should export this type to allow other plugins to read the `meta`
+ *                   field in a type-safe way.
+ */
+type ServiceStatus<Meta extends Record<string, any> = unknown> = {
+  /**
+   * The current availability level of the service.
+   */
+  level: ServiceStatusLevel.available;
```

Comment:
Shouldn't that be just `ServiceStatusLevel` ?

Reply:
This union type is to express that when level is not `available`, the `summary `field is required. This first part of the union is the case where it is not required.